sqlglot/optimizer/remove_unused_columns.py

```diff
@@ -0,0 +1,162 @@
+from collections import defaultdict
+from copy import copy
+
+from sqlglot import expressions as exp
+from sqlglot.errors import OptimizeError
+from sqlglot.optimizer.helper import select_scope
+
+# Sentinel value that means an outer query selecting ALL columns
+SELECT_ALL = "*"
+
+
+def remove_unused_columns(expression):
+    """
+    Rewrite sqlglot AST to remove unused columns projections.
+
+    Example:
+        >>> import sqlglot
+        >>> sql = "SELECT y.a AS a FROM (SELECT x.a AS a, x.b AS b FROM x) AS y"
+        >>> expression = sqlglot.parse_one(sql)
+        >>> remove_unused_columns(expression).sql()
+        'SELECT y.a AS a FROM (SELECT x.a AS a FROM x) AS y'
+
+    Args:
+        expression (sqlglot.Expression): expression to optimize
+    Returns:
+        sqlglot.Expression: optimized expression
+    """
+    expression = expression.copy()
+    _pushdown_statement(expression, SELECT_ALL)
+    return expression
+
+
+def _pushdown_statement(expression, parent_selections):
+    """
+    Search SELECT or UNION for columns that can be removed.
+
+    Args:
+        expression (exp.Select or exp.Union): expression to search
+        parent_selections (set or str): columns being selected by an outer query.
+            This can be the special value `SELECT_ALL`, which mean the outer query
+            is selecting everything.
+    Returns:
+        dict: Mapping of selectable names to columns.
+        This is used during recursion, so the outer query can:
+            1. pullup any selected columns
+            2. pushdown selected columns into CTEs
+    """
+    if isinstance(expression, exp.Select):
+        return _pushdown_select(expression, parent_selections)
+    if isinstance(expression, exp.Union):
+        if expression.args.get("distinct"):
+            # We can't remove selections on UNION ALL
```

Comment:
is this comment reversed? if it's distinct it means it's not a union all

Reply:
Ah, indeed. nice catch